core/src/main/java/org/elasticsearch/action/admin/indices/shards/TransportIndicesShardStoresAction.java

```diff
@@ -213,7 +213,7 @@ void finish() {
              * A shard exists/existed in a node only if shard state file exists in the node
              */
             private boolean shardExistsInNode(final NodeGatewayStartedShards response) {
-                return response.storeException() != null || response.version() != -1;
+                return response.storeException() != null || response.version() != -1 || response.allocationId() != null;
```

Comment:
Note that the condition just requires any of the fields to have a value, so it will work fine in the case where we do not have allocation ids. We say that the shard exists if we get any sensible information about it. At the moment we have a version whenever we have an allocation id. But if we would remove versions in the future, that might not hold anymore. Adding `response.allocationId() != null` just adds a little more safety.